RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFMultiDepthClusterizerHelper.h

```diff
@@ -0,0 +1,332 @@
+
+      // send last lane value (total tile offset) to lane idx = low_lane_idx:
+      const warp::warp_mask_t active_mask = 1 | get_lane_mask(high_lane_idx);
+      const unsigned int tmp = warp::shfl_mask(acc, active_mask, local_offset, high_lane_idx, w_extent);
+
+      if (lane_idx == 0)
+        local_offset = tmp;  //lane 0 keeps full (inclusive for the last lane) sum
+    }
+    return lane_idx == 0 ? local_offset : local_offset - val;  //we return exclusive sum!
+  }
+
+  /**
+ * @brief Returns logical index for a given physical lane index based on custom lane mask.
+ *
+ * @tparam TAcc Alpaka accelerator type.
+ * 
+ * @param acc Alpaka accelerator instance.
+ * @param mask Input bitmask.
+ * @param lane_idx imput phys. lane index
+ * 
+ * @return Index of the lane in the mask 
+ */
+
+  template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE unsigned int get_logical_lane_idx(TAcc const& acc,
+                                                                        const warp::warp_mask_t mask,
+                                                                        const unsigned int lane_idx) {
+    if (lane_idx == 0)
+      return lane_idx;  // nothing to do, phys idx coincide with the logical one.
+    const warp::warp_mask_t lane_mask = mask & (get_lane_mask(lane_idx) - 1);
+    return alpaka::popcount(acc, lane_mask);  // Count 1s below current lane
+  }
+
+  /**
+ * @brief Returns physical lane index for a given logical lane index based on custom lane mask.
+ *
+ * @tparam TAcc Alpaka accelerator type.
+ * 
+ * @param acc Alpaka accelerator instance.
+ * @param mask Input mask.
+ * @param logical_lane_idx input logical lane index
+ * 
+ * @return Physical index of the lane in the mask 
+ */
+
+  template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+  ALPAKA_FN_ACC ALPAKA_FN_INLINE unsigned int get_physical_lane_idx(TAcc const& acc,
+                                                                    const warp::warp_mask_t mask,
+                                                                    int logical_lane_idx) {
+    using signed_warp_mask_t = std::conditional_t<sizeof(warp::warp_mask_t) == 8, std::int64_t, std::int32_t>;
+
+    signed_warp_mask_t m = mask;
+
+    while (logical_lane_idx--)
+      m &= (m - 1);
+
+    const auto pos = alpaka::ffs(acc, m);
+
+    return static_cast<unsigned int>(pos - 1);
+  }
+
+  /**
+ * @brief generic warp reduction
+ *
+ * @tparam TAcc Alpaka accelerator type.
+ * 
+ * @param acc Alpaka accelerator instance.
+ * @param in input value to reduce
+ * @param f reducer 
+ * 
+ * @return return reduced value (propagated to all lanes in the mask by default)
+ */
+
+  template <typename TAcc,
+            typename reduce_t,
+            typename reducer_t,
+            bool all = true,
+            typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    requires std::is_arithmetic_v<reduce_t>
```

Comment:
Since `requires` is already used, this could be
```suggestion
            bool all = true>
    requires std::is_arithmetic_v<reduce_t> and alpaka::isAccelerator<TAcc>
```